File: jambeth/jambeth-util/src/main/java/de/osthus/ambeth/collections/WeakHashMap.java
package de.osthus.ambeth.collections;

import java.lang.ref.Reference;
import java.lang.ref.ReferenceQueue;

public class WeakHashMap<K, V> extends AbstractHashMap<Reference<K>, K, V>
{
	public static <K, V> WeakHashMap<K, V> create(int size)
	{
		return create(size, DEFAULT_LOAD_FACTOR);
	}

	public static <K, V> WeakHashMap<K, V> create(int size, float loadFactor)
	{
		return new WeakHashMap<K, V>((int) (size / loadFactor) + 1, loadFactor);
	}

	protected int size;

	protected final ReferenceQueue<Object> referenceQueue = new ReferenceQueue<Object>();

	public WeakHashMap()
	{
		this(DEFAULT_INITIAL_CAPACITY, DEFAULT_LOAD_FACTOR, WeakMapEntry.class);
	}

	public WeakHashMap(float loadFactor)
	{
		this(DEFAULT_INITIAL_CAPACITY, loadFactor, WeakMapEntry.class);
	}

	public WeakHashMap(int initialCapacity)
	{
		this(initialCapacity, DEFAULT_LOAD_FACTOR, WeakMapEntry.class);
	}

	public WeakHashMap(int initialCapacity, float loadFactor)
	{
		this(initialCapacity, loadFactor, WeakMapEntry.class);
	}

	public WeakHashMap(int initialCapacity, float loadFactor, Class<?> entryClass)
	{
		super(initialCapacity, loadFactor, entryClass);
	}

	@SuppressWarnings("unchecked")
	public Reference<K> getWeakReferenceEntry(K key)
	{
		final int hash = hash(extractHash(key));
		IMapEntry<K, V>[] table = this.table;
		final int i = hash & (table.length - 1);
		IMapEntry<K, V> entry = table[i];
		while (entry != null)
		{
			if (equalKeys(key, entry))
			{
				return (Reference<K>) entry;
			}
			entry = entry.getNextEntry();
		}
		return null;
	}

	@Override
	protected IMapEntry<K, V> createEntry(int hash, K key, V value, IMapEntry<K, V> nextEntry)
	{
		return new WeakMapEntry<K, V>(key, value, hash, nextEntry, referenceQueue);
	}

	@Override
	protected void entryAdded(final IMapEntry<K, V> entry)
	{
		size++;
		checkForCleanup();
	}

	@Override
	protected final void entryRemoved(IMapEntry<K, V> entry)
	{
		entryRemovedNoCleanup(entry);
		checkForCleanup();
	}

	protected void entryRemovedNoCleanup(IMapEntry<K, V> entry)
	{
		size--;
	}

	@Override
	protected void setNextEntry(IMapEntry<K, V> entry, IMapEntry<K, V> nextEntry)
	{
		((WeakMapEntry<K, V>) entry).setNextEntry(nextEntry);
	}

	@Override
	public int size()
	{
		return size;
	}

	@SuppressWarnings("unchecked")
	public void checkForCleanup()
	{
		IMapEntry<K, V>[] table = this.table;
		int tableLengthMinusOne = table.length - 1;
		IMapEntry<K, V> removedEntry;
		ReferenceQueue<Object> referenceQueue = this.referenceQueue;
		while ((removedEntry = (WeakMapEntry<K, V>) referenceQueue.poll()) != null)
		{
			int i = removedEntry.getHash() & tableLengthMinusOne;
			IMapEntry<K, V> entry = table[i];
			if (entry == null)
			{
				// Nothing to do
				continue;
			}
			if (entry == removedEntry)
			{
				table[i] = entry.getNextEntry();
				entryRemovedNoCleanup(entry);
				continue;
			}
			IMapEntry<K, V> prevEntry = entry;
			entry = entry.getNextEntry();
			while (entry != null)
			{
				if (entry == removedEntry)
				{
					setNextEntry(prevEntry, entry.getNextEntry());
					entryRemovedNoCleanup(entry);
					break;
				}
				prevEntry = entry;
				entry = entry.getNextEntry();
			}
		}
	}
}

File: ambeth/Ambeth.Util/ambeth/collections/WeakHashMap.cs
using System;
using System.Collections.Generic;

namespace De.Osthus.Ambeth.Collections
{
    public class WeakHashMap<K, V> : AbstractHashMap<IMapEntry<K, V>, K, V>
    {
        public static WeakHashMap<K, V> Create(int size)
        {
            return Create(size, DEFAULT_LOAD_FACTOR);
        }

        public static WeakHashMap<K, V> Create(int size, float loadFactor)
        {
            return new WeakHashMap<K, V>((int)(size / loadFactor) + 1, loadFactor);
        }

        protected int size;

        public WeakHashMap()
            : this(DEFAULT_INITIAL_CAPACITY, DEFAULT_LOAD_FACTOR, typeof(WeakMapEntry<K, V>))
        {
            // intended blank
        }

        public WeakHashMap(float loadFactor)
            : this(DEFAULT_INITIAL_CAPACITY, loadFactor, typeof(WeakMapEntry<K, V>))
        {
            // intended blank
        }

        public WeakHashMap(int initialCapacity)
            : this(initialCapacity, DEFAULT_LOAD_FACTOR, typeof(WeakMapEntry<K, V>))
        {
            // intended blank
        }

        public WeakHashMap(int initialCapacity, float loadFactor)
            : this(initialCapacity, loadFactor, typeof(WeakMapEntry<K, V>))
        {
            // intended blank
        }

        public WeakHashMap(int initialCapacity, float loadFactor, Type entryClass)
            : base(initialCapacity, loadFactor)
        {
            // intended blank
        }

        public WeakReference GetWeakReferenceEntry(K key)
        {
            int hash = Hash(ExtractHash(key));
            IMapEntry<K, V>[] table = this.table;
            int i = hash & (table.Length - 1);
            IMapEntry<K, V> entry = table[i];
            while (entry != null)
            {
                if (EqualKeys(key, entry))
                {
                    return (WeakReference)entry;
                }
                entry = entry.NextEntry;
            }
            return null;
        }

        protected override IMapEntry<K, V> CreateEntry(int hash, K key, V value, IMapEntry<K, V> nextEntry)
        {
            return new WeakMapEntry<K, V>(key, value, hash, nextEntry);
        }

        protected override void EntryAdded(IMapEntry<K, V> entry)
        {
            size++;
        }

        protected override void EntryRemoved(IMapEntry<K, V> entry)
        {
            size--;
        }

        protected override void SetNextEntry(IMapEntry<K, V> entry, IMapEntry<K, V> nextEntry)
        {
            ((WeakMapEntry<K, V>)entry).SetNextEntry(nextEntry);
        }

        protected override V SetValueForEntry(IMapEntry<K, V> entry, V value)
        {
            return ((WeakMapEntry<K, V>)entry).SetValue(value);
        }

        public override int Count
        {
            get
            {
                return size;
            }
        }

        protected override void Transfer(IMapEntry<K, V>[] newTable)
        {
            int newCapacityMinus1 = newTable.Length - 1;
            IMapEntry<K,V>[] table = this.table;

            for (int a = table.Length; a-- > 0; )
            {
                IMapEntry<K, V> entry = table[a], next;
                while (entry != null)
                {
                    next = GetNextEntry(entry);

                    // only handle this entry if it has still a valid key
                    if (entry.Key != null)
                    {
                        int i = entry.Hash & newCapacityMinus1;
                        SetNextEntry(entry, newTable[i]);
                        newTable[i] = entry;
                    }
                    entry = next;
                }
            }
            base.Transfer(newTable);
        }
    }
}